src/test/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActorTest.scala

```diff
@@ -56,6 +57,63 @@ class AppTaskLauncherActorTest extends MarathonSpec with GivenWhenThen {
     Mockito.verify(taskTracker).get(app.id)
   }
 
+  test("Upgrading an app updates app definition in actor and requeries backoff") {
+    Given("an entry for an app")
+    Mockito.when(taskTracker.get(app.id)).thenReturn(Set(marathonTask))
+    val launcherRef = createLauncherRef(instances = 0)
+    rateLimiterActor.expectMsg(RateLimiterActor.GetDelay(app))
+    rateLimiterActor.reply(RateLimiterActor.DelayUpdate(app, clock.now()))
+    val counts = Await.result(launcherRef ? AppTaskLauncherActor.GetCount, 3.seconds).asInstanceOf[QueuedTaskCount]
+    assert(counts.tasksLaunchedOrRunning == 1)
+
+    When("upgrading the app")
+    val upgradedApp = app.copy(cmd = Some("new command"))
+    launcherRef ! AppTaskLauncherActor.AddTasks(upgradedApp, 1)
+
+    Then("the actor requeries the backoff delay")
+    rateLimiterActor.expectMsg(RateLimiterActor.GetDelay(upgradedApp))
+    val newDelay: Timestamp = clock.now() + 5.seconds
+    rateLimiterActor.reply(RateLimiterActor.DelayUpdate(upgradedApp, newDelay))
+    val counts2 = Await.result(launcherRef ? AppTaskLauncherActor.GetCount, 3.seconds).asInstanceOf[QueuedTaskCount]
+    assert(counts2.backOffUntil == newDelay)
+
+    And("the actor knows the new app definition")
+    assert(counts2.app == upgradedApp)
+    assert(counts2.tasksLaunchedOrRunning == 1)
+
+    // We don't care about these:
+    Mockito.reset(taskTracker)
+  }
+
+  test("Upgrading an app updates reregisters the offerMatcher at the manager") {
+    Given("an entry for an app")
+    Mockito.when(taskTracker.get(app.id)).thenReturn(Set(marathonTask))
+    val launcherRef = createLauncherRef(instances = 1)
+    rateLimiterActor.expectMsg(RateLimiterActor.GetDelay(app))
+    rateLimiterActor.reply(RateLimiterActor.DelayUpdate(app, clock.now()))
+    val counts = Await.result(launcherRef ? AppTaskLauncherActor.GetCount, 3.seconds).asInstanceOf[QueuedTaskCount]
+    assert(counts.tasksLaunchedOrRunning == 1)
+
+    // We don't care about interactions until this point
+    Mockito.reset(offerMatcherManager)
+
+    When("upgrading the app")
+    val upgradedApp = app.copy(cmd = Some("new command"))
+    launcherRef ! AppTaskLauncherActor.AddTasks(upgradedApp, 1)
+    rateLimiterActor.expectMsg(RateLimiterActor.GetDelay(upgradedApp))
+    rateLimiterActor.reply(RateLimiterActor.DelayUpdate(upgradedApp, clock.now()))
+
+    // wait for message being processed
+    Await.result(launcherRef ? AppTaskLauncherActor.GetCount, 3.seconds).asInstanceOf[QueuedTaskCount]
+
+    Then("the actor reregisters itself for at the offerMatcher")
+    Mockito.verify(offerMatcherManager).removeSubscription(mockito.Matchers.any())(mockito.Matchers.any())
```

Comment:
it would be even better if we also verify the order, using inOrder:

http://mockito.googlecode.com/hg-history/1.6/javadoc/org/mockito/InOrder.html